corehq/apps/locations/adjacencylist.py

```diff
@@ -0,0 +1,174 @@
+from __future__ import absolute_import
+from django.db.models import IntegerField
+from django.db.models.expressions import Value
+from django.db.models.query import Q, QuerySet
+from django_cte import With
+from mptt.models import MPTTModel, TreeManager
+
+from .queryutil import ComparedQuerySet, TimingContext
+
+int_field = IntegerField()
+
+
+class AdjListManager(TreeManager):
+
+    def cte_get_ancestors(self, node, ascending=False, include_self=False):
+        """Query node ancestors
+
+        :param node: A model instance or a QuerySet or Q object querying
+        the adjacency list model. If a QuerySet, it should query a
+        single value with something like `.values('pk')`. If Q the
+        `include_self` argument will be ignored.
+        :param ascending: Order of results. The default (`False`) gets
+        results in descending order (root ancestor first, immediate
+        parent last).
+        :param include_self:
+        :returns: A `QuerySet` instance.
+        """
+        parent_col = self.model.parent_id_attr
+
+        if isinstance(node, Q):
+            where = node
+        elif include_self:
+            if isinstance(node, QuerySet):
+                where = Q(pk__in=node)
+            else:
+                where = Q(pk=node.id)
+        elif isinstance(node, QuerySet):
+            where = Q(pk__in=node.values(parent_col))
```

Comment:
how long can the list be of values to check with `IN`?

you may already know this but queries with `IN` clauses that have long lists perform slowly

Reply:
Yes, that's true. This will be a subquery, not a list. Something like
```sql
SELECT ...
FROM locations_sqllocation
WHERE id IN (SELECT parent_id FROM locations_sqllocation WHERE ...)
```

Reply:
ok, that means that less sql will go over the wire for long lists but the performance of the query itself still won't be as good as if it used a join with an indexed column to do this kind of filtering.

Reply:
True. The focus of this effort is to introduce CTE queries with as little impact on existing code in order to improve the performance of location uploads. Currently the (MPTT) code would execute this subquery and then construct a new queryset with the results from that. I think eliminating DB round-trips will result in an overall performance improvement here.

If we found this to be a bottleneck we could optimize those places where it is a problem to use a join on an indexed column, as you suggest.